user/tlu/module/src/TluRawEvent2StdEventConverter.cc

```diff
@@ -29,41 +29,142 @@ bool TluRawEvent2StdEventConverter::Converting(eudaq::EventSPC d1, eudaq::Standa
+      finets_vec.emplace_back(finets3 & 0xFF);
+  }
+  if ((triggersFired >> 4) & 0x1) {
+      finets_vec.emplace_back(finets4 & 0xFF);
+  }
+  if ((triggersFired >> 5) & 0x1) {
+      finets_vec.emplace_back(finets5 & 0xFF);
+  }
+
+  // Now average over all vector elements:
+  // Initialize to 0th vector element:
+  uint32_t finets_avg = 0xFF & finets_vec.at(0);
+
+  // Calculate "rolling average":
+  for(int i=1; i<finets_vec.size(); i++) {
+      // 128 = half the fineTS counter range -> 128*780ps = 99ns = 3m (TOF)
+      // For a time difference larger than 99ns, the overflow detection doesn't work anymore.
+      if(abs(finets_avg - finets_vec.at(i)) > 128) { // 128*780ps = 99ns
+          finets_vec.at(i) += 0xFF; // overflow compensation
+      }
+      // Need to weight average with previous number of iterations:
+      // (Don't truncate to 8-bit for correct averaging, i.e. don't apply 0xFF)
+      finets_avg = ((i*finets_avg + finets_vec.at(i)) / (i+1));
+
+      // This leads to rounding errors on the 781ps binning level
+      // but otherwise the overflow detection logic breaks down.
+  }
 
+  auto coarse_ts = static_cast<uint64_t>(d1->GetTimestampBegin() / 25);
+
+  uint8_t coarse_3lsb = 0x3 & coarse_ts;
+  uint8_t finets_avg_3msb = (0xE0 & finets_avg) >> 5;
+
+  // Casting to 32 bit is essential to detect the overflow in the 8th bit!
+  // 0x100 = 200ns (range of 3 bits with 25ns binning)
+  if((static_cast<uint32_t>(coarse_3lsb) - finets_avg_3msb) > 0x100) {
+      // coarse_ts is delayed by 2 clock cycles -> roll back
+      coarse_ts -= 2;
+  }
+
+  // with combined fineTS: replace the 3lsb of the coarse timestamp with the 3msb of the finets_avg:
+  auto ts = static_cast<uint64_t>((25. / 32. * (((coarse_ts << 5) & 0xFFFFFFFFFFFFFF00) + (finets_avg & 0xFF))) * 1000.);
+
+  // Set times for StdEvent in picoseconds (timestamps provided in nanoseconds):
   d2->SetTimeBegin(ts);
   d2->SetTimeEnd(d1->GetTimestampEnd() * 1000);
   d2->SetTimestamp(ts, d1->GetTimestampEnd(), d1->IsFlagTimestamp());
 
   // Identify the detetor type
   d2->SetDetectorType("TLU");
-  d2->SetTag("TRIGGER", triggersFired);
-  d2->SetTag("FINE_TS0", finets0);
-  d2->SetTag("FINE_TS1", finets1);
-  d2->SetTag("FINE_TS2", finets2);
-  d2->SetTag("FINE_TS3", finets3);
-  d2->SetTag("FINE_TS4", finets4);
-  d2->SetTag("FINE_TS5", finets5);
+  d2->SetTag("TRIGGER", d1->GetTag("TRIGGER" , "0"));
+
+  // forward original tags:
+  d2->SetTag("FINE_TS0", d1->GetTag("FINE_TS0", "0")); // forward original tag
+  d2->SetTag("FINE_TS1", d1->GetTag("FINE_TS1", "0")); // forward original tag
+  d2->SetTag("FINE_TS2", d1->GetTag("FINE_TS2", "0")); // forward original tag
+  d2->SetTag("FINE_TS3", d1->GetTag("FINE_TS3", "0")); // forward original tag
+  d2->SetTag("FINE_TS4", d1->GetTag("FINE_TS4", "0")); // forward original tag
+  d2->SetTag("FINE_TS5", d1->GetTag("FINE_TS5", "0")); // forward original tag
+
+  // calculate (delayed) fine timestamps in ns:
```

Comment:
I'm not sure if all this information is really necessary... Maybe we can make this optional? Let's see what @eyiliu 's opinion on that topic is.

Reply:
@eyiliu - comments?